spec/services/ad_renewal_letters_export_service_spec.rb

```diff
@@ -0,0 +1,46 @@
+# frozen_string_literal: true
+
+require "rails_helper"
+
+RSpec.describe AdRenewalLettersExportService do
+  describe ".run" do
+    let(:bucket) { double(:bucket) }
+    let(:result) { double(:result, successful?: true) }
+
+    context "when there are expiring registrations from AD users" do
+      before do
+        create_list(:registration, 3, :ad_registration)
+        allow(WasteExemptionsBackOffice::Application.config).to receive(:ad_letters_exports_expires_in).and_return(35)
+        WasteExemptionsEngine::RegistrationExemption.update_all(expires_on: 35.days.from_now)
+        expect(DefraRuby::Aws).to receive(:get_bucket).and_return(bucket)
+        expect(bucket).to receive(:load).and_return(result)
+      end
+
+      it "load a file to a AWS bucket and record the content created" do
+        expect(Airbrake).to_not receive(:notify)
+
+        expect { described_class.run }.to change { WasteExemptionsEngine::AdRenewalLettersExport.count }.by(1)
+      end
+
+      context "when one registration is in a broken status and a PDF cannot be generated for it" do
+        it "raises an error on Airbrake but continue the generation for the other letters" do
```

Comment:
Well, the complication of doing that is that we need the ability to read PDF content which we currently don't have. Since the test is the same as the above one (same setup) and I don't `raise` any other error as I test it outside of this context, then I am 99% sure the error occuring is coming from the extra `registration` I am creating in this scenario which is in a broken status (no addresses). Open on suggestions on how to make it clearer :)

Reply:
I think the issue with just checking Airbrake gets notified is that it would happen if one letter fails, or if the whole job failed.

However if the whole job were to fail, I'm assuming the `AdRenewalLettersExport` record would not get created. Whereas if we got the notification because just one letter failed, then the record would still be created.

Tie this into only recording the number of letters that were exported successfully, we could possibly add this to the test

```ruby
expect { described_class.run }.to change { WasteExemptionsEngine::AdRenewalLettersExport.count }.by(1)
expect(WasteExemptionsEngine::AdRenewalLettersExport.last.number_of_letters).to eq(3)
```

Reply:
Sounds good :)

Reply:
You will have already probably spotted this, but as we're no longer capturing the number that got printed, there is no point in having the second line in my example. Just checking a record got created would be enough.